### PR TITLE
Generalize classifiers from TF / Keras /  PyTorch / MXNet to other types of models than neural networks

### LINKED ISSUE #50: Adapt ART to other types of models than neural networks
This mostly implies changing the way different backend implementations handle model outputs. In order to be effective, most attacks need to be applied at logits level. Some of the new models types ART will support do not have logits. The following describes the proposed solution for this problem:

- All backend `__init__` functions would have a flag specifying if the output of the given model represents logits. When such a model is created, attacks are applied on the regular model output, as was done previously.
- If the outputs are not logits, we can check if the sum of the outputs on a fake input is 1. In that case, we have detected probabilities as outputs and will try to estimate the logits from them. These would not be the real logits of the model, but would perform numerical scaling that makes attacks efficient.
- If the outputs are not logits and don't sum to 1 either, this might be a different type of model, and attacks will be applied directly on its outputs.
- Another flag could indicate that the outputs of the model are probabilities, but the user explicitly requests that the previous layer (logits) should be used for attacks. This option is specific to the frameworks that support the notion of layer. Moreover, finding the correct logits would not be guaranteed.
- The loss function for a given model will also be part of the input provided by the user. The loss will be applied directly on the output of the model.

In practice, the implications are that predictions / gradient computations will not be available at both logits and probabilities level, as it is currently the case. This would attract a change in the `Classifier` API for the respective functions.

### PR DESCRIPTION
**Pull request by ririnicolae**
# Description

- These changes impact TF / Keras / PyTorch / MXNet classifiers.
- They now all accept custom / general loss functions taken as input parameters.
- Parameter `logits=True/False` has been removed from `Classifier` API in functions `predict` and `class_gradient`. The predictions / gradients are now computed at the output point of the model.
- All attacks and other methods have been updated to the changes of the interface.
- Examples / notebooks have been updated.
- Classifiers' docstrings have been updated to inform users that logit outputs are more effective for attack success.
- One issue was discovered during this extension: because model predictions can be negative, KL divergence computation in VAT sometimes yields `nan`s. I will open a separate issue for this.
- See also #50 for additional comments.

Fixes #50, #75

## Type of change

Please check all relevant options.

- [ ] Bug fix (non-breaking)
- [X] New feature (non-breaking)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Testing

- [X] Unit tests

# Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by ririnicolae** [comment]
@beat-buesser Thanks for merging! To answer your questions:

- Yes, the commented test in VAT is the one that fails as described in #120.
- When the `logits` parameter was part of the `class_gradient` function signature, the `DetectorClassifier` had 8 cases to test (4 with `logits=True` and 4 with `False`). Without the parameter, only 4 cases remain to be tests, thus the change.
